hub/hub.go

```diff
@@ -4,80 +4,198 @@ import (
+		o.debug = true
+	}
 }
 
-// NewHub creates a hub using the Viper configuration.
-func NewHub(v *viper.Viper) (*Hub, error) {
-	if err := ValidateConfig(v); err != nil {
-		return nil, err
+// WithDemo enables the demo mode.
+func WithDemo() Option {
+	return func(h *Hub) {
+		h.demo = true
 	}
+}
 
-	var (
-		logger Logger
-		err    error
-	)
+// WithMetrics enables collection of Prometheus metrics.
+func WithMetrics() Option {
+	return func(h *Hub) {
+		h.metrics = NewMetrics()
+	}
+}
 
-	if v.GetBool("debug") {
-		logger, err = zap.NewDevelopment()
-	} else {
-		logger, err = zap.NewProduction()
+// WithSubscriptions allows to dispatch updates when subscriptions are created or terminated.
+func WithSubscriptions() Option {
+	return func(h *Hub) {
+		h.subscriptions = true
 	}
-	if err != nil {
-		return nil, fmt.Errorf("can't initialize zap logger: %w", err)
+}
+
+// WithLogger sets the logger to use.
+func WithLogger(logger Logger) Option {
+	return func(h *Hub) {
+		h.logger = logger
 	}
+}
 
-	t, err := NewTransport(v, logger)
-	if err != nil {
-		return nil, err
+// WithWriteTimeout sets maximum duration before closing the connection, defaults to 600s, set to 0 to disable.
+func WithWriteTimeout(timeout time.Duration) Option {
+	return func(h *Hub) {
+		h.writeTimeout = timeout
+	}
+}
+
+// WithDispatchTimeout sets the maximum dispatch duration of an update.
+func WithDispatchTimeout(timeout time.Duration) Option {
+	return func(h *Hub) {
+		h.dispatchTimeout = timeout
+	}
+}
+
+// WithHeartbeat enables heartbeat.
+func WithHeartbeat(interval time.Duration) Option {
+	return func(h *Hub) {
+		h.heartbeat = interval
+	}
+}
+
+// WithPublisherJWTConfig sets the JWT key and algorithm to use.
+func WithPublisherJWTConfig(key []byte, signingMethod jwt.SigningMethod) Option {
+	return func(h *Hub) {
+		h.publisherJWTConfig = &jwtConfig{key, signingMethod}
+	}
+}
+
+// WithSubscriberJWTConfig sets the JWT key and algorithm to use.
+func WithSubscriberJWTConfig(key []byte, signingMethod jwt.SigningMethod) Option {
+	return func(h *Hub) {
+		h.subscriberJWTConfig = &jwtConfig{key, signingMethod}
 	}
+}
 
-	return NewHubWithTransport(v, t, logger, NewTopicSelectorStore()), nil
+// WithPublishOrigins sets the origins allowed to publish updates.
+func WithPublishOrigins(origins []string) Option {
+	return func(h *Hub) {
+		h.publishOrigins = origins
+	}
 }
 
-// NewHubWithTransport creates a hub.
-func NewHubWithTransport(v *viper.Viper, t Transport, logger Logger, tss *TopicSelectorStore) *Hub {
-	return &Hub{
-		v,
-		logger,
-		t,
-		nil,
-		nil,
-		tss,
-		NewMetrics(),
+// WithTransport sets the transport to use.
+func WithTransport(t Transport) Option {
+	return func(h *Hub) {
+		h.transport = t
 	}
 }
 
-// Start is an helper method to start the Mercure Hub.
-func Start() {
-	h, err := NewHub(viper.GetViper())
+// WithTransportURL sets the transport to use by parsing the provided URL.
+func WithTransportURL(tu string, l Logger) Option {
```

Comment:
I think the Hub should accept a transport directly. When using it as a library, the transport URL might have a different form.

Reply:
I was working on this while you were reviewing, sorry. Could you check the latest version? I totally refactored this part.

Reply:
Sure!